Document tagged addresses (#10744)
hashicorp/consul: website/content/docs/discovery/services.mdx

##### Supported tags

* `lan` - The IPv4 LAN address at which the node or service is accessible.
* `lan_ipv4` - The IPv4 LAN address at which the node or service is accessible.
* `lan_ipv6` - The IPv6 LAN address at which the node or service is accessible.

<CodeTabs heading="Example LAN tagged address configuration">

<CodeBlockConfig filename="redis-service.hcl">

```hcl
service {
  name = "redis"
  address = "192.0.2.10"
  port = 80
  tagged_addresses {
    lan = {
      address = "192.0.2.10"
      port = 80
    }
    lan_ipv4 = {
      address = "192.0.2.10"
      port = 80
    }
    lan_ipv6 = {
      address = "2001:db8:1:2:cafe::1337"
      port = 80
    }
  }
}
```

</CodeBlockConfig>

<CodeBlockConfig filename="redis-service.json">

```json
{
  "service": {
    "name": "redis",
    "address": "192.0.2.10",
    "port": 80,
    "tagged_addresses": {
      "lan": {
        "address": "192.0.2.10",
        "port": 80
      },
      "lan_ipv4": {
        "address": "192.0.2.10",
        "port": 80
      },
      "lan_ipv6": {
        "address": "2001:db8:1:2:cafe::1337",
        "port": 80
      }
    }
  }
}
```

</CodeBlockConfig>
</CodeTabs>

#### Virtual

> Virtual tagged addresses are a logical address type which is used by

Reviewer: ```suggestion
Virtual tagged addresses are logical address types used by
```

[Connect](/docs/connect) to route traffic to the associated destination service

> in the service mesh when:

Reviewer: Just to be clear, both conditions should be met to route traffic--not either or, correct? If so, consider making it more explicit.

Author: Yes, both conditions are required to be met. I can add such phrasing to the previous sentence.

1. [Transparent proxy](/docs/connect/transparent-proxy) is enabled.
1. The service is not configured for individual instances to be
   [dialed directly](/docs/connect/config-entries/service-defaults#dialeddirectly).

-> **Note**: Virtual addresses are not required to be routable IPs within the
network. They are strictly a control plane construct used to provide a single
address for the instances of a given logical service.

##### Supported tags

> The following address tag can be configured for a service to advertise the
> logical address to be used to dial it over the service mesh.

Reviewer: ```suggestion
Use the following address tag to specify the logical address the service should advertise when dialing over the service mesh.
```
I am not certain that I correctly restated the part about "dialing over the service mesh".

Author: ```suggestion
Use the following address tag to specify the logical address at which the service can be reached by other services in the mesh.
```
Is this perhaps more clear?

Reviewer: Yes, I think I was thrown off by the term 'dial' -- thinking it meant something more technical than literally just a service communicating with another another service

- `virtual` - The virtual IP address at which a logical service is reachable.

<CodeTabs heading="Example virtual tagged address configuration">

<CodeBlockConfig filename="redis-service.hcl">

```hcl
service {
  name = "redis"
  address = "192.0.2.10"
  port = 80
  tagged_addresses {
    virtual = {
      address = "203.0.113.50"
      port = 80
    }
  }
}
```

</CodeBlockConfig>

<CodeBlockConfig filename="redis-service.json">

```json
{
  "service": {
    "name": "redis",
    "address": "192.0.2.10",
    "port": 80,
    "tagged_addresses": {
      "virtual": {
        "address": "203.0.113.50",
        "port": 80
      }
    }
  }
}
```

</CodeBlockConfig>
</CodeTabs>

#### WAN

These addresses are intended to be accessible only when dialing a node or service

> from a remote WAN federated data center.

Reviewer: It's probably just me, but I'm having trouble understanding the type of data center at line 455. 

Do we mean "remote WAN-federated data center", which would mean that the DC is remote and federated according to WAN addresses?

Do we mean "remote, WAN, federated data center", which means that the federated DC is also remote and identified by WAN address? 

Feel free to ignore this comment if it's just a gap in my understanding.

Author: Consul has a feature called [translated addresses](https://www.consul.io/api-docs#translated-addresses) which allow the catalog to return different addresses in the discovery response based on whether the requestor is in the same DC, or a different DC that is connected via WAN federation.

The docs for the [`translate_wan_addrs`](https://www.consul.io/docs/agent/options#translate_wan_addrs) config option offers a better explanation on this behavior.

> If set to true, Consul will prefer a node's configured WAN address when servicing DNS and HTTP requests for a node in a remote datacenter. This allows the node to be reached within its own datacenter using its local address, and reached from other datacenters using its WAN address, which is useful in hybrid setups with mixed networks. This is disabled by default.

These `wan` tagged addresses are the same as the `WAN address` mentioned in the previous paragraph.

Reviewer: OK, then I _believe_ the correct statement would be: 

"Define a WAN address when services need to remotely access a federated datacenter at a WAN address." 

Yeah?

Author: We settled on the following phrasing.
> Define a WAN address for the service or node when it should be accessed at an alternate address by services in a remote datacenter.
